### DISCUSSION TITLE
time_stats TypeError (method vs dict)

### DISCUSSION THREAD

**Discussion by walterrowe** [python-gitlab/python-gitlab#2523] [General]
I have two instances of GitLab. One is the free community edition. One is EE Premiere Edition.

I query gitlab for a list of issues and accumulate total time spent from time_stats.

I am getting different return types for time_stats between the CE and EE editions. On CE it seems to be a method. On EE it seems to be a dict (as documented in the GitLab API).

On the community edition it seems the time_stats returned with the issue is a method that returns a dict.

If I use this code against the CE (free) edition:

```python
    time_stats = this_issue.time_stats['total_time_spent']` and run against the CE edition I get this error
````

I get this error:

```text
  File "/Users/wrowe/git-repos/python-gitlab/./time-spent.py", line 209, in <module>
    time_stats = this_issue.time_stats['total_time_spent']
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
TypeError: 'method' object is not subscriptable
```

I changed my code as follows and tested it against both editions:

```python
    if type(this_issue.time_stats) is dict:
        time_stats = this_issue.time_stats
    else:
        time_stats = this_issue.time_stats()
```

This works around the issue. Is this expected?

The GitLab API indicates time_stats should be a dict at all times.

https://docs.gitlab.com/ee/api/issues.html

```json
      "time_stats": {
         "time_estimate": 0,
         "total_time_spent": 0,
         "human_time_estimate": null,
         "human_total_time_spent": null
      },
```

**Comment by nejch** [comment]
TLDR: Just use `time_stats()` in both cases.

---

@walterrowe this if _sort_ of expected, and comes from the fact that `time_stats` is both an attribute returned by the API (but only sometimes, as you've found out), as well as its own endpoint under the issues API (https://docs.gitlab.com/ee/api/issues.html#get-time-tracking-stats). This causes a collision.

python-gitlab actually compensates for that with a custom `time_stats()` method and it will either return the existing attribute or make an extra API call if not present:

https://github.com/python-gitlab/python-gitlab/blob/9aecc9e5ae1e2e254b8a27283a0744fe6fd05fb6/gitlab/mixins.py#L738-L750

So just always use `this_issue.time_stats()` and python-gitlab will know whether it needs to make an extra API call or return the existing data.

**Reply by walterrowe** [reply]
Oh if only it were that easy ... I did as you suggested. The EE edition reports an error:

```shell
Traceback (most recent call last):
  File "time-spent.py", line 222, in <module>
    time_stats = this_issue.time_stats()
                 ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'dict' object is not callable
```

I should note that the issue object comes from a group.issues.list() or project.issues.list().  The group.issues.list includes subgroups.

Do the group list and project list issue objects differ?

I will leave my work around in place. It works.

```shell
% pip list
Package            Version
------------------ ---------
certifi            2022.12.7
charset-normalizer 3.1.0
et-xmlfile         1.1.0
idna               3.4
numpy              1.24.2
openpyxl           3.1.2
pandas             1.5.3
pip                23.0.1
Pygments           2.14.0
python-dateutil    2.8.2
python-gitlab      3.13.0
pytz               2022.7.1
PyYAML             6.0
requests           2.28.2
requests-toolbelt  0.10.1
setuptools         67.6.0
six                1.16.0
urllib3            1.26.15
wheel              0.40.0

% python --version
Python 3.11.2
```

**Reply by nejch** [reply]
Oh, I didn't realize you were working with different objects. Yes, in that case best to handle it yourself as you're essentially working with 2 different types of class instances initialized from different endpoints.

See https://python-gitlab.readthedocs.io/en/stable/faq.html#i-cannot-edit-the-merge-request-issue-i-ve-just-retrieved for other similar situations.

**Reply by walterrowe** [reply]
@nejch thanks much. that makes sense. my script lets you specify a list of groups or projects and uses the corresponding object issues.list() api. this api library is amazing. so grateful for it.
